README.md

```diff
@@ -27,6 +27,70 @@ $ IPFS_RUST_EXEC=$(pwd)/rust.sh npm test
 $ cat /tmp/rust.log
 ```
 
+# Troubleshooting hangs
+
+If the tests hang because of some deadlock on the `ipfs-http` side the tests
+will print the final summary from `mocha` and then seem to wait forever. This
+is true at least for the hangs seen so far, but you could run into a different
+issues. Note that in addition to test timeouts the http client also has a
+timeout, which can keep the `npm test` alive for less than ten seconds after
+the summary has been printed.
+
+There is [`why-is-node-running`] which is a tool which can dump all the reasons
+for node process staying alive. These outputs haven't really been that useful
+in the real debugging though. Some stacktraces are missing and it can take a
+while to understand what the stacktraces represent and once you do understand,
+the understanding will not help you make progress.
+
+What has worked previously is:
+
+ 1. disable tests until you find the one which causes the hang
+ 2. rerun tests using the `rust.sh` wrapper which gives you logs at `/tmp/rust.log`
+ 3. continue debugging over at `ipfs-http`
+
+"Disable tests" as in, comment tests out or run them with `IPFS_RUST_EXEC=...
+npm test -- --grep 'should XYZ'`. You should get to a single running test which
+hangs. If the test does a lot, you can refactor that into a smaller one which
+will only cause the hang and nothing else.
+
+The use of `rust.sh` wrapper is critical as it'll give you the logging output.
+If the test has multiple running instances you might be better of separating
+logs into files per invocation of `rust.sh` by appending `.$$` to the log files
+name, which will expand to the process id of the shell running the script.
+
+"Continue debugging" is a more tricky one. What has worked previously is:
+
+ 1. attaching a `gdb` to a running process in the case of livelocks
+ 2. adding debugging in the case of "everything stalled"
+
+(I might be using the term wrong here) Livelocks happen when a task (running on
+either tokio or async-std) never returns from [`std::future::Future::poll`] but
+stays busy all of the time. You can spot these by seeing a core being
+utilitized by `ipfs-http` constantly. These are easy to track down by:
+
+ 1. attach `gdb -p $process_id path/to/your/ipfs-http`
```

Comment:
This is all useful and I think we can include this here for now, but I'm wondering if we should include a debugging / troubleshooting guide in rust-ipfs proper in the end.

Reply:
My idea for adding it here was that I debugged a hang which required that "shutdown while having request open" which I didn't figure out to do otherwise but yeah we need to figure out a better place. I just worry that under `rust-ipfs/http/docs` or something this will never catch anyone's eyes.